### PR TITLE
[#23578] Towards supporting JDK9 j.u.c.Flow interfaces

### PR DESCRIPTION
**Pull request by ktoso**
🔥 🔥 🔥  **Work in Progress** - this does not work yet as it needs the plugin ecosystem to be bumped and JDK9 on our Jenkins🔥 🔥 🔥 

Preparing for the "Future of Java" ™ 😉

### CODE REVIEW

**Comment by ktoso** [comment]
This resolves https://github.com/akka/akka/issues/23578 once done

File: project/plugins.sbt [hunk 1]
```diff
@@ -1,45 +1,30 @@
...
+addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.1.0-M1")
+addSbtPlugin("com.eed3si9n" % "sbt-unidoc" % "0.4.1")
+addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings" % "2.0.0")
+addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.27")
+//addSbtPlugin("pl.project13.sbt" % "sbt-jol" % "0.1.3")
+addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.2.2-RC2")
+addSbtPlugin("io.spray" % "sbt-boilerplate" % "0.6.1")
+addSbtPlugin("com.timushev.sbt" % "sbt-updates" % "0.3.1")
+//addSbtPlugin("com.lightbend.akka" % "sbt-paradox-akka" % "0.3")
+//addSbtPlugin("com.lightbend" % "sbt-whitesource" % "0.1.5")
```
Comment (raboof) [inline project/plugins.sbt]: 0.1.6 is released with sbt 1.0, but it looks like we cannot use that yet due to https://github.com/typesafehub/sbt-whitesource/issues/20

File: akka-stream/src/main/java-jdk9/akka/stream/javadsl/Jdk9Sink.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com/>
+ */
+
+package akka.stream.javadsl;
+
+/**
+ * TODO java doc
+ */
+public interface Jdk9Sink {
```
Comment (raboof) [inline akka-stream/src/main/java-jdk9/akka/stream/javadsl/Jdk9Sink.java]: Would the name `Jdk9Sink` get confusing once JDK 10 is out?

Comment (ktoso) [reply akka-stream/src/main/java-jdk9/akka/stream/javadsl/Jdk9Sink.java]: Yeah, names are pretty bad still... I'm torn between `JavaFlowSinkSupport` or something similar...

File: project/plugins.sbt (hunk 1 above)
Comment (ktoso) [reply project/plugins.sbt]: Thanks for the info 👍

**Comment by ktoso** [comment]
Eh conflicts. Hoped I addressed those... :(

Please review though. Due to the biuld changes some formatting may be changed so review with whitespace insensitive diff  (like this ?https://github.com/akka/akka/pull/23579/files?w=1 )

File: akka-stream-tests/src/test/java-jdk9-only/akka/stream/javadsl/JavaFlowSupportCompileTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      JavaFlowSupport.Flow.fromProcessor(() -> processor);
+    final akka.stream.javadsl.Flow<String, String, NotUsed> stringStringNotUsedFlow1 = 
+      JavaFlowSupport.Flow.fromProcessorMat(() -> Pair.apply(processor, NotUsed.getInstance()));
+
+    final Sink<String, Flow.Publisher<String>> stringPublisherSink = 
+      JavaFlowSupport.Sink.asPublisher(AsPublisher.WITH_FANOUT);
+    final Sink<String, NotUsed> stringNotUsedSink = 
+      JavaFlowSupport.Sink.fromSubscriber(processor);
+  }
+}
```
Comment (ktoso) [inline akka-stream-tests/src/test/java-jdk9-only/akka/stream/javadsl/JavaFlowSupportCompileTest.java]: A compile spec

File: akka-stream/src/main/scala-jdk9-only/akka/stream/scaladsl/JavaFlowSupport.scala [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+import scala.annotation.unchecked.uncheckedVariance
+
+/**
+ * For use only with `JDK 9+`.
+ *
+ * Provides support for `java.util.concurrent.Flow.*` interfaces which mirror the Reactive Streams
+ * interfaces from `org.reactivestreams`. See: [http://www.reactive-streams.org/](reactive-streams.org).
+ */
+object JavaFlowSupport {
```
Comment (ktoso) [inline akka-stream/src/main/scala-jdk9-only/akka/stream/scaladsl/JavaFlowSupport.scala]: this is the APIs for Scala

File: akka-stream/src/main/scala-jdk9-only/akka/stream/scaladsl/JavaFlowSupport.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com/>
+ */
+
+package akka.stream.scaladsl
```
Comment (ktoso) [inline akka-stream/src/main/scala-jdk9-only/akka/stream/scaladsl/JavaFlowSupport.scala]: notice the directory, it's `akka-stream/src/main/scala-jdk9-only/akka/stream/scaladsl/JavaFlowSupport.scala`
which is different than the `scala-jdk8` which style I had reserved for the multi-release jars... But feel free to suggest better names

File: akka-stream/src/main/scala/akka/stream/FanOutShape.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright (C) 2014-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream
+
+import scala.collection.immutable
+import scala.annotation.unchecked.uncheckedVariance
+
+object FanOutShape {
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/FanOutShape.scala]: Rebasing was very weird... Didn't @jrudolph extract those into separate files now?

File: akka-stream/README.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+Directory structure explained
+=============================
+
+Please note that the `src/main/java-jdk9-compile` are NOT for multi-release-jar support,
+these are directories that are included in the build only if running on JDK9.
+This is a MUST for releasing Akka including `j.u.c.Flow` support (numerous alternatives were explored).
```
Comment (ktoso) [inline akka-stream/README.md]: FYI

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/StageActorRefSpec.scala [hunk 8]
```diff
@@ -193,8 +193,8 @@ object StageActorRefSpec {
 
         def behaviour(m: (ActorRef, Any)): Unit = {
           m match {
-            case (sender, Add(n))                ⇒ sum += n
-            case (sender, PullNow)               ⇒ pull(in)
+            case (sender, Add(n)) ⇒ sum += n
+            case (sender, PullNow) ⇒ pull(in)
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/StageActorRefSpec.scala]: damn formatting :(

File: akka-stream/src/main/java-jdk9-only/akka/stream/javadsl/JavaFlowSupport.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+import akka.japi.Creator;
+import akka.stream.impl.JavaFlowAndRsConverters;
+
+/**
+ * For use only with `JDK 9+`.
+ * <p>
+ * Provides support for `java.util.concurrent.Flow.*` interfaces which mirror the Reactive Streams
+ * interfaces from `org.reactivestreams`. See <http//www.reactive-streams.org/>(reactive-streams.org).
+ */
+public final class JavaFlowSupport {
```
Comment (ktoso) [inline akka-stream/src/main/java-jdk9-only/akka/stream/javadsl/JavaFlowSupport.java]: The JavaDSL meat

File: akka-stream/src/main/scala-jdk9-only/akka/stream/impl/JavaFlowAndRsConverters.scala [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+ * Please note that either of these types are designed for *inter-op* and usually should not be used directly 
+ * in applications. The intended use case is for shared libraries, like database drivers or similar to provide
+ * the inter-operable types, such that other librarie can co-operate with them directly, if that is your use case
+ * and you're using the j.u.c.Flow types, use the [[akka.stream.scaladsl.JavaFlowSupport]] sources/sinks/flows instead.
+ *
+ * The interfaces as well as semantic contract of these sets of interfaces.
+ * Please see https://github.com/reactive-streams/reactive-streams-jvm to learn more about the specification,
+ * and TCK that is available to verify an implementation is a valid implementation.
+ */
+object JavaFlowAndRsConverters {
```
Comment (ktoso) [inline akka-stream/src/main/scala-jdk9-only/akka/stream/impl/JavaFlowAndRsConverters.scala]: The actual converters

File: akka-stream/src/main/scala-jdk9-only/akka/stream/impl/JavaFlowAndRsConverters.scala [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    }
+  final def asRs[T](s: Flow.Subscriber[T]): rs.Subscriber[T] = s match {
+    case null ⇒ null // null remains null
+    case adapter: RsSubscriberToJavaFlowAdapter[T] ⇒ adapter.delegate // unwrap adapter instead of wrapping again
+    case _ =>  new JavaFlowSubscriberToRsAdapter[T](s)
+  }
+
+  final def asJava[T, R](p: rs.Processor[T, R]): Flow.Processor[T, R] = p match {
+      case null ⇒ null // null remains null
+      case adapter: JavaFlowProcessorToRsAdapter[T, R] ⇒ adapter.delegate // unwrap adapter instead of wrapping again
```
Comment (ktoso) [inline akka-stream/src/main/scala-jdk9-only/akka/stream/impl/JavaFlowAndRsConverters.scala]: same behaviour as the ones that may ship from RS eventually

**Comment by ktoso** [comment]
Killing this one in favour of "just the classes", but this one is very useful since includes the actual build changes to get it to actually build (basically disabling plugins)

File: akka-stream/src/main/scala/akka/stream/FanOutShape.scala (hunk 6 above)
Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/FanOutShape.scala]: Yes.

**Pull request closed without merging**